tests/test_clean.py

```diff
@@ -1,96 +1,210 @@
+import os
+
 from html5lib.filters.base import Filter
 import pytest
 
-import bleach
+from bleach import clean
 from bleach.sanitizer import Cleaner
 
 
-def test_empty():
-    assert bleach.clean('') == ''
+def test_clean_idempotent():
+    """Make sure that applying the filter twice doesn't change anything."""
+    dirty = '<span>invalid & </span> < extra http://link.com<em>'
+    assert clean(clean(dirty)) == clean(dirty)
 
 
-def test_nbsp():
-    assert bleach.clean('&nbsp;test string&nbsp;') == '&nbsp;test string&nbsp;'
+def test_only_text_is_cleaned():
+    some_text = 'text'
+    some_type = int
+    no_type = None
 
+    assert clean(some_text) == some_text
 
-def test_comments_only():
-    comment = '<!-- this is a comment -->'
-    assert bleach.clean(comment) == ''
-    assert bleach.clean(comment, strip_comments=False) == comment
+    with pytest.raises(TypeError) as e:
+        clean(some_type)
+    assert "argument cannot be of 'type' type" in str(e)
 
-    open_comment = '<!-- this is an open comment'
-    assert bleach.clean(open_comment) == ''
-    assert (
-        bleach.clean(open_comment, strip_comments=False) ==
-        '{0!s}-->'.format(open_comment)
-    )
+    with pytest.raises(TypeError) as e:
+        clean(no_type)
+    assert "NoneType" in str(e)
 
 
-def test_with_comments():
-    text = '<!-- comment -->Just text'
-    assert bleach.clean(text) == 'Just text'
-    assert bleach.clean(text, strip_comments=False) == text
+def test_empty():
+    assert clean('') == ''
 
 
-def test_no_html():
-    assert bleach.clean('no html string') == 'no html string'
+def test_content_has_no_html():
+    assert clean('no html string') == 'no html string'
 
 
-def test_allowed_html():
-    assert (
-        bleach.clean('an <strong>allowed</strong> tag') ==
+@pytest.mark.parametrize('data, expected', [
+    (
+        'an <strong>allowed</strong> tag',
         'an <strong>allowed</strong> tag'
-    )
-    assert (
-        bleach.clean('another <em>good</em> tag') ==
+    ),
+
+    (
+        'another <em>good</em> tag',
         'another <em>good</em> tag'
     )
+])
+def test_content_has_allowed_html(data, expected):
+    assert clean(data) == expected
 
 
-def test_bad_html():
+def test_html_is_lowercased():
     assert (
-        bleach.clean('a <em>fixed tag') ==
-        'a <em>fixed tag</em>'
+        clean('<A HREF="http://example.com">foo</A>') ==
+        '<a href="http://example.com">foo</a>'
```

Comment:
I moved this test from the bottom of the file, improved the function name, and changed the test to use allowed-by-default things to reduce the complexity.

Reply:
👍